firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,293 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                } else {
+                    <u8 as PartialOrd>::gt
+                };
+
+                let children = get_children_iter(branch)
+                    .filter(move |(_, child_pos)| compare_op(child_pos, &pos));
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+
+            New(NodeType::Branch(branch)) => {
+                if let Some(value) = branch.value.as_ref() {
+                    let value = value.to_vec();
+                    return Ok(Some((node.into_node(), value)));
+                }
+
+                let children = get_children_iter(branch);
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+        }
+
+        first_loop = false;
     }
 }
 
-/// Compute a key from a set of parents
-fn key_from_parents(parents: &[(ObjRef, u8)]) -> Vec<u8> {
-    parents[1..]
-        .chunks_exact(2)
-        .map(|parents| (parents[0].1 << 4) + parents[1].1)
-        .collect::<Vec<u8>>()
+fn get_children_iter(branch: &BranchNode) -> impl Iterator<Item = (DiskAddress, u8)> {
+    branch
+        .children
+        .into_iter()
+        .enumerate()
+        .filter_map(|(pos, child_addr)| child_addr.map(|child_addr| (child_addr, pos as u8)))
+}
+
+#[must_use]
+struct MustUse<T>(T);
+
+impl<T> From<T> for MustUse<T> {
+    fn from(t: T) -> Self {
+        Self(t)
+    }
 }
-fn key_from_parents_and_leaf(parents: &[(ObjRef, u8)], leaf: &LeafNode) -> Vec<u8> {
-    let mut iter = parents[1..]
+
+impl<T: Not> Not for MustUse<T> {
+    type Output = T::Output;
+
+    fn not(self) -> Self::Output {
+        self.0.not()
+    }
+}
+
+fn next_child<'a, S, T, Iter>(
+    merkle: &'a Merkle<S, T>,
+    mut children: Iter,
+    parents: &mut Vec<(ObjRef<'a>, u8)>,
+    node: &mut ParentNode<'a>,
+    pos: &mut u8,
+) -> Result<MustUse<bool>, super::MerkleError>
+where
+    Iter: Iterator<Item = (DiskAddress, u8)>,
+    S: ShaleStore<Node>,
+{
+    if let Some((child_addr, child_pos)) = children.next() {
+        let child = merkle.get_node(child_addr)?;
+
+        *pos = child_pos;
+        let node = std::mem::replace(node, ParentNode::New(child));
+        parents.push((node.into_node(), *pos));
+    } else {
+        let Some((next_parent, next_pos)) = parents.pop() else {
+            return Ok(false.into());
+        };
+
+        *node = ParentNode::Visited(next_parent);
+        *pos = next_pos;
+    }
+
+    Ok(true.into())
+}
+
+/// create an iterator over the key-nibbles from all parents _excluding_ the sentinal node.
+fn nibble_iter_from_parents<'a>(parents: &'a [(ObjRef, u8)]) -> impl Iterator<Item = u8> + 'a {
```

Comment:
I think this does a lot of allocations, since you are creating a vec for each component as you traverse. I think it's possible to return a `impl Iterator<Item = &'a u8>` with something like this (importing the `either` crate here, but there are other ways):

```
fn nibble_iter_from_parents<'a>(parents: &'a [(ObjRef, u8)]) -> impl Iterator<Item = &'a u8> {
    parents
        .iter()
        .skip(1) // always skip the sentinal node
        .flat_map(|(parent, child_nibble)| match parent.inner() {
            NodeType::Branch(_) => either::Either::Left(std::iter::once(child_nibble)),
            NodeType::Extension(extension) => either::Either::Right(extension.path.iter()),
            NodeType::Leaf(leaf) => either::Either::Right(leaf.path.iter()),
        })
}
```

There is then some work to always use references in the callers, of course, but that actually seemed a bit easier as you're using `copied()` to match this in.

Reply:
`either` is a good idea, will try it out.

Reply:
Why did you decide not to go with an iterator over `&'a u8` as suggested above?

Reply:
> Why did you decide not to go with an iterator over `&'a u8` as suggested above?

Answering my own question: it's actually faster since `size_of::<u8>` < `size_of::<&u8>`

Reply:
Yeah, I think I generally prefer to copy early if I'm going to need to copy (not clone), even if it was a `u64`

Reply:
Great suggestion, thanks again